hail/python/hail/linalg/blanczos_data_script.py

```diff
@@ -0,0 +1,265 @@
+    return result_arr_sum
+
+def computeNextH(A, H):
+    nextG = matmul_colblocked_rowblocked(A, H)
+    return matmul_rowblocked_nonblocked(A, nextG)
+
+def hailBlanczos(A, G, m, n, k, l, q, block_size):
+    
+    # assert l > k
+    # assert (q+1)*l <= (n - k)
+    # assert n <= m
+    
+    start = time.time()
+    
+    Hi = matmul_rowblocked_nonblocked(A, G)
+    npH = concatToNumpy(Hi)
+    for j in range(0, q):
+        Hj = computeNextH(A, Hi)
+        npH = np.concatenate((npH, concatToNumpy(Hj)), axis=1)
+        Hi = Hj
+    
+    # assert npH.shape == (m, (q+1)*l)
+    # perform QR decomposition on unblocked version of H
+    Q, R = np.linalg.qr(npH)
+    # assert Q.shape == (m, (q+1)*l)
+    # block Q's rows into the same number of blocks that A has
+    blocked_Q_table = ndarray_to_table(chunk_ndarray(Q, block_size))
+    
+    T = matmul_colblocked_rowblocked(blocked_Q_table, A)
+    # assert T.shape == ((q+1)*l, n)
+    
+    U, S, W = np.linalg.svd(T, full_matrices=False)
+    # assert U.shape == ((q+1)*l, n)
+    # assert S.shape == (n,)
+    # assert W.shape == (n, n)
+    
+    sing_val = S[k]
+    
+    V = matmul_rowblocked_nonblocked(blocked_Q_table, U)
+    arr_V = concatToNumpy(V)
+    
+    end = time.time()
+    
+    truncV = arr_V[:,:k]
+    truncS = S[:k]
+    truncW = W[:k,:]
+        
+    return truncV, truncS, truncW, sing_val, Q, end - start #, bound, satC
+
+
+def timeNumpyLibrary(data):
+    start = time.time()
+    _, S, _ = np.linalg.svd(data, full_matrices=False)
+    end = time.time()
+    return end - start
+
+def makeSharedData(model_input, block_size):
+    
+    # we should have m > n for hail implementation
+    mt = hl.balding_nichols_model(*model_input)
+    
+    mt.write("balding_nichols_data.mt", overwrite=True)
+    mt = hl.read_matrix_table("balding_nichols_data.mt")
+    
+    mt = mt.transmute_entries(n_alt = hl.float64(mt.GT.n_alt_alleles())) 
+    table = matrix_table_to_table_of_ndarrays(mt.n_alt, block_size, tmp_path='/tmp/test_table.ht')
+    
+    # for numpy implementation we want transposed version so m < n
+    np_matrix = np.asmatrix(concatToNumpy(table))
+
+    return table, np_matrix
+
+
+
+# SCRIPT:
+
+
+df = pd.DataFrame(columns=['M', 'N', 'block size', 'K', 'L', 'Q', 'blanczos time', 'numpy time'])
+
+# references a dataframe df not passed into the function
+def loop(i, m, n, block_size, k, l, q):
+    
+    print("loop attempt ", i, ' with m:', m, ', n:', n, ', m*n =', m*n, ', and q:', q)
+
+    try: 
+        assert l > k
+        assert n <= m
+    except Exception as e:
+        print(e)
+        return
+
+    try:
+        table, mat = makeSharedData((3, n, m), block_size)
+    except Exception as e:
+        print(e)
+        print('failed to make data with ', (m, n))
+        return
+    
+    try:
+        G = hl.nd.array(np.random.normal(0, 1, (n,l)))
+        _, S, _, _, _, hail_time  = hailBlanczos(table, G, m, n, k, l, q, block_size)
```

Comment:
Is there a reason you don't just have this script call the new `hl._blanczos_pca`?

Reply:
I started writing it before I had the tests for the `_blanczos_pca` method working, I can change this now if I want. Though, I have timers and other useful data collection in this script, so I might want to keep it as well, if I don't want to change the return types of the hail method.